model/registry/db.yaml

```diff
@@ -529,3 +533,67 @@ groups:
           reference the [elasticsearch schema](https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json)
           in order to map the path part values to their names.
         examples: ['db.elasticsearch.path_parts.index=test-index', 'db.elasticsearch.path_parts.doc_id=123']
+  - id: registry.db.vector
+    prefix: db.vector
+    type: attribute_group
+    brief: >
+      This group defines attributes for vector databases.
+    attributes:
+      - id: similarity
+        type:
+          allow_custom_values: true
+          members:
+            - id: cosine
+              value: 'cosine'
+              brief: >
+                The cosine metric.
+              stability: experimental
+            - id: dot
+              value: 'dot'
+              brief: >
+                The dot product metric.
+              stability: experimental
+            - id: euclidean
+              value: 'euclidean'
+              brief: >
+                The euclidean distance metric.
+              stability: experimental    
+            - id: manhattan
+              value: 'manhattan'
+              brief: >
+                The Manhattan distance metric.
+              stability: experimental
+        stability: experimental
+        brief: >
+          The metric used in similarity search.
+        examples: 'cosine'
+      - id: id
+        type: string
+        stability: experimental
+        brief: >
+          The ID of vector.
+        examples: '5c56c793-69f3-4fbf-87e6-c4bf54c28c26'
+      - id: name
+        type: string
+        stability: experimental
+        brief: >
+          The name field as of the vector (e.g. a field name).
```

Comment:
do we need both - the name and the id?

Reply:
The `id` is the identifier of the vector, the `name` is the field name that contains the vector. Many database uses both but some uses only `name` of the field. Maybe we should use a better naming here.

Reply:
@lmolkova what do you think about this? Thanks.

Reply:
Could you provide some examples of id and name? 

Let's pick a few (2-3) popular databases and explain what id and name would mean in their context.

E.g. I look into [MongoDB](https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-type/#about-the-vector-type) and I don't understand what id or name would be.

Or I look into [Azure Search](https://learn.microsoft.com/en-us/azure/search/vector-store) and don't understand what should go into id.

I look into [pinecone](https://docs.pinecone.io/guides/get-started/quickstart#5-create-a-serverless-index) and it does not talk about ids.

Also, perhaps by vector you mean index?

Reply:
@lmolkova Here some examples:
- MongoDB uses name for the field (i.e. `path` in search [here](https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-stage/#ann-examples) and `field-name` in the definition [here]( https://www.mongodb.com/docs/atlas/atlas-search/field-types/knn-vector/#define-the-index-for-the-fts-field-type-type))
- Azure Search uses name ([here](https://learn.microsoft.com/en-us/azure/search/vector-search-how-to-create-index?tabs=config-2024-07-01%2Crest-2024-07-01%2Cpush%2Cportal-check-index#add-a-vector-field-to-the-fields-collection))
- Pinecone uses id ([here](https://docs.pinecone.io/guides/get-started/quickstart#6-upsert-vectors))
- Qdrant uses id (Point ID, [here](https://qdrant.tech/documentation/concepts/points/#point-ids))
- Elasticsearch uses name ([here](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html))
- Milvus uses id ([here](https://milvus.io/docs/insert-update-delete.md#Insert-entities))
- Chroma uses id ([here](https://docs.trychroma.com/getting-started#4.-add-some-text-documents-to-the-collection))
- pgvector for PostgreSQL uses name (i.e. the field name with type vector(x), [here](https://github.com/pgvector/pgvector?tab=readme-ov-file#getting-started))
- Redis uses name (i.e. the field name that will contain the vector values in  `fieldname_embedding`, [here](https://redis.io/docs/latest/develop/get-started/vector-database/#4-generate-text-embeddings))

By **db.vector.name** I mean the **field name** in the document that will be used for embedding (most common for DB like MongoDB, Azure Search, Redis, Elasticsearch, PostgreSQL) and by **db.vector.id** I mean the identifier of the vector (most used in native vector db like Pinecone, Qdrant, Chroma, Milvus). 

Basically these attributes, **name** and **ID**, are used to identify the vector, maybe a better naming for `db.vector.name` can be `db.vector.field_name`?

Reply:
Thank you for the context!

> name

I agree that `db.vector.field_name` would be more descriptive.
It could be difficult to collect it though - for cosmos, postgres and many other dbs it would require parsing queries (and specific parsing for vector search). I.e. it'd probably mean that very few generic DB instrumentations will do it. 

Even the native instrumentation that we have in CosmosDB does not know if query supplied by user is doing vector search. 

So I wonder how critical this attribute is for the observability purposes. Also I'd be mostly static - is it important enough to justify additional costs of populating it on each span?

A  typical semconv decision making hint is: if we're not sure we need it, let's not add it. Adding attributes to spans is easy, removing them is breaking and hard.

Reply:
> id

I still don't understand what's behind `db.vector.id` - it seems to be a generic record id and there is nothing vector-specific here. 

I support adding generic attribute like `db.record.id: string` or `db.record.ids: string[]` (needs polishing)

Reply:
The `db.vector.id` is the record id but specific for the vector. The `db.record.id` will work as well, since it's just an identifier for the record (i.e. vector).

I think we should add `db.vector.field_name` since it is used in many database to specify which field has been used for the embedding.